### PR TITLE
[Shell] Avoid providing new lines when code segment contains diagnostics

### LINKED ISSUE #34539: [Shell] Ballerina shell provides incorrect new lines when function body contains error diagnostics
**Description:**

Consider following example

Screenshot from 2022-01-06 14-17-27

### PR DESCRIPTION
**Pull request by DamithaSenevirathne**
Fixes #34539

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ExpressionValidator.java [hunk 1]
```diff
@@ -42,20 +43,15 @@ public void setNextValidator(Validator nextValidator) {
...
         }
 
-        if (!isIncomplete) {
-            return nextInValidator.evaluate(source);
-        } else {
+        if (!parsedNode.apply(incompleteInputFinder)) {
             return true;
         }
 
+        return nextInValidator.evaluate(source);
```
Comment (nadeeshaan) [inline ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ExpressionValidator.java:55]: We can merge these together

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ModuleMemberValidator.java [hunk 2]
```diff
@@ -51,25 +50,20 @@ public boolean evaluate(String source) {
         TextDocument document = TextDocuments.from(source);
         SyntaxTree tree = SyntaxTree.from(document);
         ModulePartNode node = tree.rootNode();
-        NodeList<ModuleMemberDeclarationNode> members = node.members();
-        if (members.size() > 1) {
-            return false;
+        // Ignore code segments including imports
+        if (node.imports().size() > 0) {
```
Comment (nadeeshaan) [inline ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ModuleMemberValidator.java:54]: What if we use the `isEmpty()` API here?

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ModuleMemberValidator.java [hunk 3]
```diff
@@ -51,25 +50,20 @@ public boolean evaluate(String source) {
...
         }
 
-        if (!isIncomplete) {
-            return  nextInValidator.evaluate(source);
-        } else {
+        if (!parsedNode.apply(incompleteInputFinder)) {
             return true;
         }
 
+        return nextInValidator.evaluate(source);
```
Comment (nadeeshaan) [inline ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ModuleMemberValidator.java:67]: We can merge these together

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/StatementValidator.java [hunk 4]
```diff
@@ -28,13 +29,29 @@
...
+        Node parsedNode = NodeParser.parseBlockStatement("{" + source + "}");
+        if (!parsedNode.hasDiagnostics()) {
+            return true;
+        }
+
+        if (!parsedNode.apply(incompleteInputFinder)) {
+            return true;
+        }
+
+        return nextInValidator.evaluate(source);
```
Comment (nadeeshaan) [inline ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/StatementValidator.java:55]: We can merge these together

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/WorkerValidator.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+/**
+ * Validates user input as a worker statement.
+ *
+ * @since 2.0.0
+ */
+public class WorkerValidator implements Validator {
+
+    public WorkerValidator() {
+    }
```
Comment (nadeeshaan) [inline ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/WorkerValidator.java:33]: Do we need this default constructor?

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/utils/IncompleteInputFinder.java [hunk 6]
```diff
@@ -47,17 +71,40 @@
...
     @Override
     public Boolean transform(BinaryExpressionNode node) {
         return node.rhsExpr().apply(this);
     }
 
+    @Override
+    public Boolean transform(FunctionDefinitionNode node) {
+        if (node.functionKeyword().isMissing()) {
+            return true;
+        }
```
Comment (nadeeshaan) [inline ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/utils/IncompleteInputFinder.java:95]: What if we combine this with line 97? Same applies for the other transform methods as well

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/utils/IncompleteInputFinder.java [hunk 7]
```diff
@@ -82,28 +129,98 @@ public Boolean transform(ListConstructorExpressionNode node) {
...
+
+    @Override
+    public Boolean transform(ModuleVariableDeclarationNode node) {
+        // Diagnostic check to avoid case : annotation not attached to construct
+        for (Diagnostic diagnostic : StreamSupport.stream(node.diagnostics().spliterator(), false)
+                .collect(Collectors.toList())) {
+            if (diagnostic.diagnosticInfo().code().equals("BCE0524")) {
+                return false;
+            }
         }
```
Comment (nadeeshaan) [inline ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/utils/IncompleteInputFinder.java:185]: Try to use the stream APIs

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ExpressionValidator.java (hunk 1 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ExpressionValidator.java]: Updated in 7d25e1dea188151fc8c5f84a1bf9156ce22ebc81

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ModuleMemberValidator.java (hunk 2 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ModuleMemberValidator.java]: Updated in 7d25e1dea188151fc8c5f84a1bf9156ce22ebc81

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ModuleMemberValidator.java (hunk 3 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ModuleMemberValidator.java]: Updated in 7d25e1dea188151fc8c5f84a1bf9156ce22ebc81

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/StatementValidator.java (hunk 4 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/StatementValidator.java]: Updated in 7d25e1dea188151fc8c5f84a1bf9156ce22ebc81

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/WorkerValidator.java (hunk 5 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/WorkerValidator.java]: Removed in  7d25e1dea188151fc8c5f84a1bf9156ce22ebc81

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/utils/IncompleteInputFinder.java (hunk 6 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/utils/IncompleteInputFinder.java]: Updated in 7d25e1dea188151fc8c5f84a1bf9156ce22ebc81. Updated other possible cases as well.

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/utils/IncompleteInputFinder.java (hunk 7 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/utils/IncompleteInputFinder.java]: Updated in 7d25e1dea188151fc8c5f84a1bf9156ce22ebc81

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ExpressionValidator.java [hunk 8]
```diff
@@ -42,20 +43,11 @@ public void setNextValidator(Validator nextValidator) {
...
-
-        if (!isIncomplete) {
-            return nextInValidator.evaluate(source);
-        } else {
+        Node parsedNode = NodeParser.parseExpression(source);
+        if (!parsedNode.hasDiagnostics() || !parsedNode.apply(incompleteInputFinder)) {
             return true;
         }
 
+        return nextInValidator.evaluate(source);
```
Comment (nadeeshaan) [inline ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ExpressionValidator.java:51]: Can't we use `return !parsedNode.hasDiagnostics() || !parsedNode.apply(incompleteInputFinder) || nextInValidator.evaluate(source)`

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ModuleMemberValidator.java [hunk 9]
```diff
@@ -51,25 +50,12 @@ public boolean evaluate(String source) {
...
-        if (!isIncomplete) {
-            return  nextInValidator.evaluate(source);
-        } else {
+        // Ignore code segments including imports
+        Node parsedNode = NodeParser.parseModuleMemberDeclaration(source);
+        if (!node.imports().isEmpty() || !parsedNode.hasDiagnostics() || !parsedNode.apply(incompleteInputFinder)) {
             return true;
         }
 
+        return nextInValidator.evaluate(source);
```
Comment (nadeeshaan) [inline ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ModuleMemberValidator.java:59]: Can't we use `return node.imports().isEmpty() || !parsedNode.hasDiagnostics() || !parsedNode.apply(incompleteInputFinder) || nextInValidator.evaluate(source)`?
Applies for the other cases as well

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ExpressionValidator.java (hunk 8 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/validator/ExpressionValidator.java]: Updated in ea2b94a0c6e70886dde225592eb0985699ae6232

**Review by nadeeshaan — DISMISSED** [review]
LGTM

**Pull request merged**